### PR TITLE
Tn20 redo the repository separating action compositions from db.runs

### PR DESCRIPTION
**Pull request by joaov2**

### CODE REVIEW

File: app/repositories/slick/implementations/SlickChatsRepository.scala [hunk 1]
```diff
@@ -93,57 +93,75 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
     } yield (chatId, emailId, body, date, sent)
   //endregion
 
+  def cleanRepeatedDateEmails(chatPreviewSeq: Seq[ChatPreview]): Seq[ChatPreview] = {
+    val indexed = chatPreviewSeq.zipWithIndex
```
Comment (IvoSaraiva) [inline app/repositories/slick/implementations/SlickChatsRepository.scala]: Try `chatPreviewSeq.zip(chatPreviewSeq.tail)`

File: app/repositories/slick/implementations/SlickChatsRepository.scala [hunk 2]
```diff
@@ -93,57 +93,75 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
...
+   *         The preview of each chat only shows the most recent email
    */
-  def getChatsPreview(mailbox: Mailbox, userId: String): Future[Seq[ChatPreview]] = {
+  def getChatsPreviewAction(mailbox: Mailbox, userId: String) = {
 
-    val groupedQuery = getVisibleEmailsQuery(userId, Some(mailbox))
+    val visibleEmailsQuery = getVisibleEmailsQuery(userId, Some(mailbox))
+
+    val groupedQuery = visibleEmailsQuery
       .map(visibleEmailRow => (visibleEmailRow._1, visibleEmailRow._4))
```
Comment (IvoSaraiva) [inline app/repositories/slick/implementations/SlickChatsRepository.scala]: Use pattern match. The ._1 and ._4 are not very clear.

File: app/repositories/slick/implementations/SlickChatsRepository.scala [hunk 3]
```diff
@@ -93,57 +93,75 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
...
 
     val chatPreviewQuery = for {
-      (chatId, date) <- groupedQuery
+      (chatId, maxDate) <- groupedQuery
       subject <- ChatsTable.all.filter(_.chatId === chatId).map(_.subject)
-      (emailId, body) <- EmailsTable.all.filter(emailRow =>
-        emailRow.chatId === chatId && emailRow.date === date).map(emailRow =>
-        (emailRow.emailId, emailRow.body.take(PREVIEW_BODY_LENGTH)))
+      (emailId, body, date) <- EmailsTable.all.filter(emailRow =>
+        emailRow.chatId === chatId && emailRow.date === maxDate).map(emailRow =>
```
Comment (IvoSaraiva) [inline app/repositories/slick/implementations/SlickChatsRepository.scala]: Don't see the point of using dates to filter something. It smells bad.

File: app/repositories/slick/implementations/SlickChatsRepository.scala [hunk 4]
```diff
@@ -158,7 +176,20 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
...
+   * Method to get the emails and other data of a specific chat of a user
+   * @param chatId ID of the chat requested
+   * @param userId ID of the user who requested the chat
+   * @return a Chat DTO that carries
+   *         the chat's subject,
+   *         the addresses involved in the chat,
+   *         the overseers of the chat
+   *         and the emails of the chat (that the user can see)
+   */
+  def getChat(chatId: String, userId: String): Future[Option[Chat]] =
```
Comment (IvoSaraiva) [inline app/repositories/slick/implementations/SlickChatsRepository.scala]: You got several of these methods. A generic one should be enough.

File: app/repositories/slick/implementations/SlickChatsRepository.scala (hunk 3 above)
Comment (joaov2) [reply app/repositories/slick/implementations/SlickChatsRepository.scala]: I know, but I have no choice. 
Like we talked before the previous groupBy only allows me to take the chatId and date columns. 
So I am forced to use the date to make a join that gives me the information I require.

File: app/repositories/slick/implementations/SlickChatsRepository.scala [hunk 5]
```diff
@@ -93,57 +93,75 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
...
       address <- AddressesTable.all.filter(_.addressId === addressId).map(_.address)
-    } yield (chatId, subject, address, date, body)
-
-    val resultOption = db.run(chatPreviewQuery.sortBy(_._4.desc).result)
 
-    val result = resultOption.map(_.map {
-      case (chatId, subject, address, dateOption, body) =>
-        (chatId, subject, address, dateOption.getOrElse("Missing Date"), body)
-    })
+      if emailId in visibleEmailsQuery.map(_._2)
```
[the remark this answers is not in the document]
Comment (joaov2) [reply app/repositories/slick/implementations/SlickChatsRepository.scala]: Sim, funciona

**Pull request merged**